mlflow/tracking/client.py

```diff
@@ -274,6 +275,19 @@ def restore_run(self, run_id):
         """
         self.store.restore_run(run_id)
 
+    def search_runs(self, experiment_ids, filter_string, run_view_type):
```

Comment:
Would it make sense to have the client accept a `SearchFilter` object rather than the filter string, and have the caller construct the `SearchFilter`?

My thinking is that it would ultimately be good if people can build queries programatically in Python, rather than having to construct a string (see, for example, this example: https://gist.github.com/acroz/b46ff5020d65137e1ff08e8139d3095a). In this case, `SearchFilter`s would be constructable from strings (using something based on the current string parsing logic) or programatically.

Reply:
@acroz : I like that idea. But to make it easy to use, we would need to document this API. The way `SearchFilter` is written today--it is fairly complex to expose out that API. When we get a chance to remove the `anded_expressions` part in MLflow `v 1.0`, we can do a cleanup. Until then, this provides a clean mechanism for users to use a programmatic way to search their runs using a simple filter string.

Reply:
Absolutely - Implementing a system like in my snippet is a lot of work and definitely outside the scope of this PR.

For the moment I'm offering only a light suggestion to have this method take a `SearchFilter` object instead of a filter string, with callers then doing:

```python
client.search_runs([1, 3, 4], SearchFilter("param.foo = 'bar'"), ViewType.ACTIVE_ONLY)
```

instead of

```python
client.search_runs([1, 3, 4], "param.foo = 'bar'", ViewType.ACTIVE_ONLY)
```

The motivation for this is to avoid having to make breaking changes to the client API in future, if it's later agreed to implement a programmatic query builder of the type I sketched out.

Reply:
Hi @acroz : No! I understand what you intended. The issue I had was that `SearchFilter` in its current state is a class that does several things (1. parse SQL construct, 2. translates anded_expression syntax from `SearchRuns` PB, and 3. does filtering of run objects based on filter criteria). In its current state--it is not consumable as an API object, which needs to be thin and easily documentable.

Two possible ways to handle your idea without exposing the internal workings of `SearchFilter` are 

1. Define a new API-grade class say `Filter` that handles translation to `SearchFilter` internal object. We can still do this before the next release to avoid the breaking change concern.

2. Have a `SearchFilterBuilder` class that is documented and can have the logic to translate to a string based query. This would be on the lines of your example. This will provide a dual flexibility of a simple string based filter approach for simple use case and users of complex expression can use programmatic mechanisms to build this string. But this would need some thinking and design. Also, this can be structured in later and would not affect this API or have a breaking change in future.

One more stronger motivating reasons, IMO, to have a string in this python (and in future Java and R) client APIs, is that this is the same interface for MLflow UI so people can cut paste the same expression.

Reply:
@acroz was just chatting with @mparkhe - it definitely does make make sense to have a more programmatic way to build up a query for searching over runs. We'd also thought in the past about exposing access to run data e.g. as a pandas or Spark DataFrame, which could also enable this use case (& allow us to avoid having to write an expression-builder API in Python/Java/R). In the short run we might like to keep the client API simple, but we probably should figure out a plan for enabling structured queries before we release 1.0 & can no longer make breaking changes.

Reply:
Thanks for your feedback on this @mparkhe and @smurching.

I think that there's a good argument in favour of passing around a structured object as it's easier for different backends (servers) to interpret and translate into their filtering logic (i.e. they don't have to re-implement the SQL parsing logic in `SearchFilter`).

This is of relevance to ongoing work in integrating MLflow into our platform, so I'd like to flesh out these ideas a little further - would you be up for e.g. creating a channel on slack to discuss this further? I might also expand on my thoughts in my linked snippet next week.